cli/flox-rust-sdk/src/models/environment/core_environment.rs

```diff
@@ -100,10 +100,123 @@ impl<State> CoreEnvironment<State> {
         fs::read_to_string(self.manifest_path()).map_err(CoreEnvironmentError::OpenManifest)
     }
 
+    /// Return the contents of the lockfile or None if it doesn't exist
+    pub fn existing_lockfile_contents(&self) -> Result<Option<String>, CoreEnvironmentError> {
+        let lockfile_path = self.lockfile_path();
+        if let Ok(lockfile_path) = CanonicalPath::new(lockfile_path) {
+            Ok(Some(
+                fs::read_to_string(lockfile_path)
+                    .map_err(LockedManifestError::ReadLockfile)
+                    .map_err(CoreEnvironmentError::LockedManifest)?,
+            ))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Return a [LockedManifest] if the lockfile exists,
+    /// otherwise return None
+    pub fn existing_lockfile(&self) -> Result<Option<LockedManifest>, CoreEnvironmentError> {
+        let lockfile_path = self.lockfile_path();
+        if let Ok(lockfile_path) = CanonicalPath::new(lockfile_path) {
+            Ok(Some(
+                LockedManifest::read_from_file(&lockfile_path)
+                    .map_err(CoreEnvironmentError::LockedManifest)?,
+            ))
+        } else {
+            Ok(None)
+        }
+    }
+
     pub fn manifest(&self) -> Result<TypedManifest, CoreEnvironmentError> {
         toml::from_str(&self.manifest_content()?).map_err(CoreEnvironmentError::DeserializeManifest)
     }
 
+    /// Return a [LockedManifest] if the environment is already locked and has
+    /// the same manifest contents as the manifest, otherwise return None.
+    /// Error if there is a pkgdb manifest that needs to be locked.
+    fn lockfile_if_up_to_date(&self) -> Result<Option<LockedManifest>, CoreEnvironmentError> {
```

Comment:
I still think the logic even of the docs is a bit complex.
Maybe that's partially because the return type is not "linear". To elaborate:

You get an `Ok(None)` for a non existing lockfile (although finding that a lockfile does not exist cannot fail). You then get `Ok(Some(_))` if it uptodate, `Err(_)` if its not.

IMO the current impl would be clearer if the return type was `Option<Result<LockedManifest, CoreEnvironmentError>>`.
In that case the first condition (lockfile exists) is followed by the second condition (lockfile is uptodate), rather than interleaving them.

---------

*non-blocking consideration*

This still doesnt change much about the complexity of the logic.
That could be addressed by returning an intermediate Wrapper for a lockfile which then allows you to unwrap conditionally on the lockfile being up to date or not.
Splitting those conditions would be clearer in my eyes but this is a purely non blocking suggestion. To add a bit color to this i'll leave la skeleton of that idea below


```rust
fn maybe_lockfile<'a>(&'a self) -> Option<LockfileGuard<'a>> {
	 let lockfile_path = self.lockfile_path();
	 CanonicalPath::new(lockfile_path).ok().map(LockfileGuard::new)
}

struct LockfileGuard<'a> {
	_lifetime: PhantomData<'a ()>,
	lockfile_path: PathBuf
}

impl LockfileGuard<'_> {
	fn new<'a>(lockfile_path: PathBuf) -> Self<'a> {
		LockfileGuard {
			_lifetime: PhantomData,
			lockfile_path
		}
	}

	/// read and parse lockfile and require that it is uptodate with the current manifest
	fn unwrap_up_to_date(self) -> Result<LockedManifest, CoreenvironmentError> {
		...
	}

	/// read and parse lockfile. lockfile may be out of date with the manifest
	/// This seems to be a replacement for `CoreEnvironment::lockfile` but that just a conjecture
	fn unwrap -> Result<LockedManifest, CoreEnvironmentError> {
		...
	}
}
```

Reply:
> You get an `Ok(None)` for a non existing lockfile (although finding that a lockfile does not exist cannot fail). You then get `Ok(Some(_))` if it uptodate, `Err(_)` if its not.

You get an Ok(None) if it isn't up to date. You only get an Err if it isn't up to date and there's a pkgdb manifest

Reply:
Well, the docs say

> Error if there is a pkgdb manifest that needs to be locked.

I read "pkgdb manifest that needs to be locked." as "lockfile isn't up to date"...

Reply:
Would "v0 manifest that needs to be locked" be more clear? It's pkgdb as opposed to catalog manifest

Reply:
Changed to
```
    /// Error if there is a v0 manifest that needs to be locked, because
    /// locking with pkgdb is no longer supported
```
both here and for `ensure_locked`

Reply:
Ohh right, yes v0 manifest if probably clearer then!